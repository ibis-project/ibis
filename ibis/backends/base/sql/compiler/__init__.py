from __future__ import annotations

from ibis.backends.base.sql.compiler.base import DDL, DML
from ibis.backends.base.sql.compiler.query_builder import (
    Compiler,
    Difference,
    Intersection,
    Select,
    SelectBuilder,
    TableSetFormatter,
    Union,
)
from ibis.backends.base.sql.compiler.translator import ExprTranslator, QueryContext

__all__ = (
    "Compiler",
    "Select",
    "SelectBuilder",
    "Union",
    "Intersection",
    "Difference",
    "TableSetFormatter",
    "ExprTranslator",
    "QueryContext",
    "DML",
    "DDL",
)
