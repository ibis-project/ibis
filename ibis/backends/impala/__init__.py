"""Impala backend"""
import ibis.config
from ibis.backends.base import BaseBackend

# these objects are exposed in the public API and are not used in the module
from .client import (  # noqa: F401
    ImpalaClient,
    ImpalaConnection,
    ImpalaDatabase,
    ImpalaDatabaseTable,
    ImpalaTable,
)
from .hdfs import HDFS, WebHDFS, hdfs_connect
from .udf import *  # noqa: F401,F403


class Backend(BaseBackend):
    name = 'impala'
    kind = 'sql'
    client = ImpalaClient
    database_class = ImpalaDatabase
    table_class = ImpalaDatabaseTable
    table_expr_class = ImpalaTable
    HDFS = HDFS
    WebHDFS = WebHDFS

    def hdfs_connect(self, *args, **kwargs):
        return hdfs_connect(*args, **kwargs)

    def connect(
        self,
        host='localhost',
        port=21050,
        database='default',
        timeout=45,
        use_ssl=False,
        ca_cert=None,
        user=None,
        password=None,
        auth_mechanism='NOSASL',
        kerberos_service_name='impala',
        pool_size=8,
        hdfs_client=None,
    ):
        """Create an ImpalaClient for use with Ibis.

        Parameters
        ----------
        host : str, optional
            Host name of the impalad or HiveServer2 in Hive
        port : int, optional
            Impala's HiveServer2 port
        database : str, optional
            Default database when obtaining new cursors
        timeout : int, optional
            Connection timeout in seconds when communicating with HiveServer2
        use_ssl : bool, optional
            Use SSL when connecting to HiveServer2
        ca_cert : str, optional
            Local path to 3rd party CA certificate or copy of server
            certificate for self-signed certificates. If SSL is enabled, but
            this argument is ``None``, then certificate validation is skipped.
        user : str, optional
            LDAP user to authenticate
        password : str, optional
            LDAP password to authenticate
        auth_mechanism : str, optional
            {'NOSASL' <- default, 'PLAIN', 'GSSAPI', 'LDAP'}.
            Use NOSASL for non-secured Impala connections.  Use PLAIN for
            non-secured Hive clusters.  Use LDAP for LDAP authenticated
            connections.  Use GSSAPI for Kerberos-secured clusters.
        kerberos_service_name : str, optional
            Specify particular impalad service principal.

        Examples
        --------
        >>> import ibis
        >>> import os
        >>> hdfs_host = os.environ.get('IBIS_TEST_NN_HOST', 'localhost')
        >>> hdfs_port = int(os.environ.get('IBIS_TEST_NN_PORT', 50070))
        >>> impala_host = os.environ.get('IBIS_TEST_IMPALA_HOST', 'localhost')
        >>> impala_port = int(os.environ.get('IBIS_TEST_IMPALA_PORT', 21050))
        >>> hdfs = ibis.impala.hdfs_connect(host=hdfs_host, port=hdfs_port)
        >>> hdfs  # doctest: +ELLIPSIS
        <ibis.filesystems.WebHDFS object at 0x...>
        >>> client = ibis.impala.connect(
        ...     host=impala_host,
        ...     port=impala_port,
        ...     hdfs_client=hdfs,
        ... )
        >>> client  # doctest: +ELLIPSIS
        <ibis.impala.client.ImpalaClient object at 0x...>

        Returns
        -------
        ImpalaClient
        """
        params = {
            'host': host,
            'port': port,
            'database': database,
            'timeout': timeout,
            'use_ssl': use_ssl,
            'ca_cert': ca_cert,
            'user': user,
            'password': password,
            'auth_mechanism': auth_mechanism,
            'kerberos_service_name': kerberos_service_name,
        }

        con = ImpalaConnection(pool_size=pool_size, **params)
        try:
            client = ImpalaClient(
                backend=self, con=con, hdfs_client=hdfs_client
            )
        except Exception:
            con.close()
            raise
        return client

    def register_options(self):
        ibis.config.register_option(
            'temp_db',
            '__ibis_tmp',
            'Database to use for temporary tables, views. functions, etc.',
        )
        ibis.config.register_option(
            'temp_hdfs_path',
            '/tmp/ibis',
            'HDFS path for storage of temporary data',
        )
