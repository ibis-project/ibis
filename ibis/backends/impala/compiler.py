from __future__ import annotations

import ibis.expr.operations as ops
from ibis.backends.base.sql.compiler import Compiler, ExprTranslator, TableSetFormatter
from ibis.backends.base.sql.registry import binary_infix_ops, operation_registry, unary


class ImpalaTableSetFormatter(TableSetFormatter):
    def _get_join_type(self, op):
        jname = self._join_names[type(op)]

        # Impala requires this
        if not op.predicates:
            jname = self._join_names[ops.CrossJoin]

        return jname

    def _format_in_memory_table(self, op):
        if op.data:
            return super()._format_in_memory_table(op)

        schema = op.schema
        names = schema.names
        types = schema.types
        rows = [
            f"{self._translate(ops.Cast(ops.Literal(None, dtype=dtype), to=dtype))} AS {name}"
            for name, dtype in zip(map(self._quote_identifier, names), types)
        ]
        return f"(SELECT * FROM (SELECT {', '.join(rows)}) AS _ LIMIT 0)"


class ImpalaExprTranslator(ExprTranslator):
    _registry = {**operation_registry, **binary_infix_ops, ops.Hash: unary("fnv_hash")}
    _forbids_frame_clause = (
        *ExprTranslator._forbids_frame_clause,
        ops.Lag,
        ops.Lead,
        ops.FirstValue,
        ops.LastValue,
    )
    _unsupported_reductions = (
        ops.ApproxMedian,
        ops.ApproxCountDistinct,
        ops.GroupConcat,
    )
    _dialect_name = "hive"
    _quote_identifiers = True


rewrites = ImpalaExprTranslator.rewrites


@rewrites(ops.FloorDivide)
def _floor_divide(op):
    return ops.Floor(ops.Divide(op.left, op.right))


class ImpalaCompiler(Compiler):
    translator_class = ImpalaExprTranslator
    table_set_formatter_class = ImpalaTableSetFormatter
