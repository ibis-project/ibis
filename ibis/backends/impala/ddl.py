from __future__ import annotations

# Copyright 2014 Cloudera Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import json

from ibis.backends.base.sql.ddl import (
    AlterTable,
    BaseDDL,
    CreateTable,
    CreateTableWithSchema,
    DropFunction,
    format_partition,
    format_schema,
    format_tblproperties,
)
from ibis.backends.base.sql.registry import type_to_sql_string


class CreateTableParquet(CreateTable):
    def __init__(
        self,
        table_name,
        path,
        example_file=None,
        example_table=None,
        schema=None,
        external=True,
        **kwargs,
    ):
        super().__init__(
            table_name,
            external=external,
            format="parquet",
            path=path,
            **kwargs,
        )
        self.example_file = example_file
        self.example_table = example_table
        self.schema = schema

    @property
    def _pieces(self):
        if self.example_file is not None:
            yield f"LIKE PARQUET '{self.example_file}'"
        elif self.example_table is not None:
            yield f"LIKE {self.example_table}"
        elif self.schema is not None:
            yield format_schema(self.schema)
        else:
            raise NotImplementedError

        yield self._storage()
        yield self._location()


class DelimitedFormat:
    def __init__(
        self,
        path,
        delimiter=None,
        escapechar=None,
        na_rep=None,
        lineterminator=None,
    ):
        self.path = path
        self.delimiter = delimiter
        self.escapechar = escapechar
        self.lineterminator = lineterminator
        self.na_rep = na_rep

    def to_ddl(self):
        yield "ROW FORMAT DELIMITED"

        if self.delimiter is not None:
            yield f"FIELDS TERMINATED BY '{self.delimiter}'"

        if self.escapechar is not None:
            yield f"ESCAPED BY '{self.escapechar}'"

        if self.lineterminator is not None:
            yield f"LINES TERMINATED BY '{self.lineterminator}'"

        yield "STORED AS TEXTFILE"
        yield f"LOCATION '{self.path}'"

        if self.na_rep is not None:
            props = {"serialization.null.format": self.na_rep}
            yield format_tblproperties(props)


class AvroFormat:
    def __init__(self, path, avro_schema):
        self.path = path
        self.avro_schema = avro_schema

    def to_ddl(self):
        yield "STORED AS AVRO"
        yield f"LOCATION '{self.path}'"

        schema = json.dumps(self.avro_schema, indent=2, sort_keys=True)
        schema = "\n".join(x.rstrip() for x in schema.splitlines())

        props = {"avro.schema.literal": schema}
        yield format_tblproperties(props)


class ParquetFormat:
    def __init__(self, path):
        self.path = path

    def to_ddl(self):
        yield "STORED AS PARQUET"
        yield f"LOCATION '{self.path}'"


class CreateTableDelimited(CreateTableWithSchema):
    def __init__(
        self,
        table_name,
        path,
        schema,
        delimiter=None,
        escapechar=None,
        lineterminator=None,
        na_rep=None,
        external=True,
        **kwargs,
    ):
        table_format = DelimitedFormat(
            path,
            delimiter=delimiter,
            escapechar=escapechar,
            lineterminator=lineterminator,
            na_rep=na_rep,
        )
        super().__init__(table_name, schema, table_format, external=external, **kwargs)


class CreateTableAvro(CreateTable):
    def __init__(self, table_name, path, avro_schema, external=True, **kwargs):
        super().__init__(table_name, external=external, **kwargs)
        self.table_format = AvroFormat(path, avro_schema)

    @property
    def _pieces(self):
        yield "\n".join(self.table_format.to_ddl())


class LoadData(BaseDDL):
    """Generate DDL for LOAD DATA command.

    Cannot be cancelled
    """

    def __init__(
        self,
        table_name,
        path,
        database=None,
        partition=None,
        partition_schema=None,
        overwrite=False,
    ):
        self.table_name = table_name
        self.database = database
        self.path = path

        self.partition = partition
        self.partition_schema = partition_schema

        self.overwrite = overwrite

    def compile(self):
        overwrite = "OVERWRITE " if self.overwrite else ""

        if self.partition is not None:
            partition = "\n" + format_partition(self.partition, self.partition_schema)
        else:
            partition = ""

        scoped_name = self._get_scoped_name(self.table_name, self.database)
        return "LOAD DATA INPATH '{}' {}INTO TABLE {}{}".format(
            self.path, overwrite, scoped_name, partition
        )


class PartitionProperties(AlterTable):
    def __init__(
        self,
        table,
        partition,
        partition_schema,
        location=None,
        format=None,
        tbl_properties=None,
        serde_properties=None,
    ):
        super().__init__(
            table,
            location=location,
            format=format,
            tbl_properties=tbl_properties,
            serde_properties=serde_properties,
        )
        self.partition = partition
        self.partition_schema = partition_schema

    def _compile(self, cmd, property_prefix=""):
        part = format_partition(self.partition, self.partition_schema)
        if cmd:
            part = f"{cmd} {part}"

        props = self._format_properties(property_prefix)
        action = f"{self.table} {part}{props}"
        return self._wrap_command(action)


class AddPartition(PartitionProperties):
    def __init__(self, table, partition, partition_schema, location=None):
        super().__init__(table, partition, partition_schema, location=location)

    def compile(self):
        return self._compile("ADD")


class AlterPartition(PartitionProperties):
    def compile(self):
        return self._compile("", "SET ")


class DropPartition(PartitionProperties):
    def __init__(self, table, partition, partition_schema):
        super().__init__(table, partition, partition_schema)

    def compile(self):
        return self._compile("DROP")


class CacheTable(BaseDDL):
    def __init__(self, table_name, database=None, pool="default"):
        self.table_name = table_name
        self.database = database
        self.pool = pool

    def compile(self):
        scoped_name = self._get_scoped_name(self.table_name, self.database)
        return f"ALTER TABLE {scoped_name} SET CACHED IN '{self.pool}'"


class CreateFunction(BaseDDL):
    _object_type = "FUNCTION"

    def __init__(self, func, name=None, database=None):
        self.func = func
        self.name = name or func.name
        self.database = database

    def _impala_signature(self):
        scoped_name = self._get_scoped_name(self.name, self.database)
        input_sig = _impala_input_signature(self.func.inputs)
        output_sig = type_to_sql_string(self.func.output)

        return f"{scoped_name}({input_sig}) returns {output_sig}"


class CreateUDF(CreateFunction):
    def compile(self):
        create_decl = "CREATE FUNCTION"
        impala_sig = self._impala_signature()
        param_line = f"location '{self.func.lib_path}' symbol='{self.func.so_symbol}'"
        return f"{create_decl} {impala_sig} {param_line}"


class CreateUDA(CreateFunction):
    def compile(self):
        create_decl = "CREATE AGGREGATE FUNCTION"
        impala_sig = self._impala_signature()
        tokens = [f"location '{self.func.lib_path}'"]

        fn_names = (
            "init_fn",
            "update_fn",
            "merge_fn",
            "serialize_fn",
            "finalize_fn",
        )

        for fn in fn_names:
            value = getattr(self.func, fn)
            if value is not None:
                tokens.append(f"{fn}='{value}'")

        joined_tokens = "\n".join(tokens)
        return f"{create_decl} {impala_sig} {joined_tokens}"


class DropFunction(DropFunction):
    def _impala_signature(self):
        full_name = self._get_scoped_name(self.name, self.database)
        input_sig = _impala_input_signature(self.inputs)
        return f"{full_name}({input_sig})"


class ListFunction(BaseDDL):
    def __init__(self, database, like=None, aggregate=False):
        self.database = database
        self.like = like
        self.aggregate = aggregate

    def compile(self):
        statement = "SHOW "
        if self.aggregate:
            statement += "AGGREGATE "
        statement += f"FUNCTIONS IN {self.database}"
        if self.like:
            statement += f" LIKE '{self.like}'"
        return statement


def _impala_input_signature(inputs):
    # TODO: varargs '{}...'.format(val)
    return ", ".join(map(type_to_sql_string, inputs))
