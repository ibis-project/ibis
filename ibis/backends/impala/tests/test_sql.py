from __future__ import annotations

import operator

import pytest

import ibis
from ibis.backends.impala.compiler import ImpalaCompiler
from ibis.backends.impala.tests.mocks import MockImpalaConnection


@pytest.fixture(scope="module")
def con():
    return MockImpalaConnection()


@pytest.fixture
def limit_cte_extract(con):
    alltypes = con.table("functional_alltypes")
    t = alltypes.limit(100)
    t2 = t.view()
    return t.join(t2).select(t)


@pytest.mark.parametrize(
    "join_type", ["cross_join", "inner_join", "left_join", "outer_join"]
)
def test_join_no_predicates_for_impala(con, join_type, snapshot):
    t1 = con.table("star1")
    t2 = con.table("star2")

    joined = getattr(t1, join_type)(t2)[[t1]]
    result = ImpalaCompiler.to_sql(joined)
    snapshot.assert_match(result, "out.sql")


def test_limit_cte_extract(limit_cte_extract, snapshot):
    case = limit_cte_extract
    result = ImpalaCompiler.to_sql(case)
    snapshot.assert_match(result, "out.sql")


def test_nested_join_base(snapshot):
    t = ibis.table(dict(uuid="string", ts="timestamp"), name="t")
    counts = t.group_by("uuid").size()
    max_counts = counts.group_by("uuid").aggregate(max_count=lambda x: x[1].max())
    result = max_counts.left_join(counts, "uuid").select(counts)
    compiled_result = ImpalaCompiler.to_sql(result)
    snapshot.assert_match(compiled_result, "out.sql")


def test_nested_joins_single_cte(snapshot):
    t = ibis.table(dict(uuid="string", ts="timestamp"), name="t")

    counts = t.group_by("uuid").size()

    last_visit = t.group_by("uuid").aggregate(last_visit=t.ts.max())

    max_counts = counts.group_by("uuid").aggregate(max_count=counts[1].max())

    main_kw = max_counts.left_join(
        counts, ["uuid", max_counts.max_count == counts[1]]
    ).select(counts)

    result = main_kw.left_join(last_visit, "uuid").select(
        main_kw, last_visit.last_visit
    )
    compiled_result = ImpalaCompiler.to_sql(result)
    snapshot.assert_match(compiled_result, "out.sql")


def test_nested_join_multiple_ctes(snapshot):
    ratings = ibis.table(
        dict(userid="int64", movieid="int64", rating="int8", timestamp="string"),
        name="ratings",
    )
    movies = ibis.table(dict(movieid="int64", title="string"), name="movies")

    expr = ratings.timestamp.cast("timestamp")
    ratings2 = ratings["userid", "movieid", "rating", expr.name("datetime")]
    joined2 = ratings2.join(movies, ["movieid"])[ratings2, movies["title"]]
    joined3 = joined2.filter([joined2.userid == 118205, joined2.datetime.year() > 2001])
    top_user_old_movie_ids = joined3.filter(
        [joined3.userid == 118205, joined3.datetime.year() < 2009]
    )[["movieid"]]
    # projection from a filter was hiding an insidious bug, so we're disabling
    # that for now see issue #1295
    cond = joined3.movieid.isin(top_user_old_movie_ids.movieid)
    result = joined3[cond]
    compiled_result = ImpalaCompiler.to_sql(result)
    snapshot.assert_match(compiled_result, "out.sql")


def test_logically_negate_complex_boolean_expr(snapshot):
    t = ibis.table(
        [("a", "string"), ("b", "double"), ("c", "int64"), ("d", "string")],
        name="t",
    )

    def f(t):
        return t.a.isin(["foo"]) & t.c.notnull()

    expr = (~f(t)).name("tmp")
    result = ImpalaCompiler.to_sql(expr)
    snapshot.assert_match(result, "out.sql")


def test_join_with_nested_or_condition(snapshot):
    t1 = ibis.table([("a", "string"), ("b", "string")], "t")
    t2 = t1.view()

    joined = t1.join(t2, [t1.a == t2.a, (t1.a != t2.b) | (t1.b != t2.a)])
    expr = joined[t1]
    result = ImpalaCompiler.to_sql(expr)
    snapshot.assert_match(result, "out.sql")


def test_join_with_nested_xor_condition(snapshot):
    t1 = ibis.table([("a", "string"), ("b", "string")], "t")
    t2 = t1.view()

    joined = t1.join(t2, [t1.a == t2.a, (t1.a != t2.b) ^ (t1.b != t2.a)])
    expr = joined[t1]
    result = ImpalaCompiler.to_sql(expr)
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize("method", ["isnull", "notnull"])
def test_is_parens(method, snapshot):
    t = ibis.table([("a", "string"), ("b", "string")], "table")
    func = operator.methodcaller(method)
    expr = t[func(t.a) == func(t.b)]

    result = ImpalaCompiler.to_sql(expr)
    snapshot.assert_match(result, "out.sql")


def test_is_parens_identical_to(snapshot):
    t = ibis.table([("a", "string"), ("b", "string")], "table")
    expr = t[t.a.identical_to(None) == t.b.identical_to(None)]

    result = ImpalaCompiler.to_sql(expr)
    snapshot.assert_match(result, "out.sql")


def test_join_aliasing(snapshot):
    test = ibis.table(
        [("a", "int64"), ("b", "int64"), ("c", "int64")], name="test_table"
    )
    test = test.mutate(d=test.a + 20)
    test2 = test[test.d, test.c]
    idx = (test2.d / 15).cast("int64").name("idx")
    test3 = test2.group_by([test2.d, idx, test2.c]).aggregate(row_count=test2.count())
    test3_totals = test3.group_by(test3.d).aggregate(total=test3.row_count.sum())
    test4 = test3.join(test3_totals, test3.d == test3_totals.d)[
        test3, test3_totals.total
    ]
    test5 = test4[test4.row_count < test4.total / 2]
    agg = (
        test.group_by([test.d, test.b])
        .aggregate(count=test.count(), unique=test.c.nunique())
        .view()
    )
    result = agg.join(test5, agg.d == test5.d)[agg, test5.total]
    result = ImpalaCompiler.to_sql(result)
    snapshot.assert_match(result, "out.sql")


def test_multiple_filters(snapshot):
    t = ibis.table([("a", "int64"), ("b", "string")], name="t0")
    filt = t[t.a < 100]
    expr = filt[filt.a == filt.a.max()]
    result = ImpalaCompiler.to_sql(expr)
    snapshot.assert_match(result, "out.sql")


def test_multiple_filters2(snapshot):
    t = ibis.table([("a", "int64"), ("b", "string")], name="t0")
    filt = t[t.a < 100]
    expr = filt[filt.a == filt.a.max()]
    expr = expr[expr.b == "a"]
    result = ImpalaCompiler.to_sql(expr)
    snapshot.assert_match(result, "out.sql")


@pytest.fixture
def region():
    return ibis.table(
        dict(r_regionkey="int16", r_name="string", r_comment="string"),
        name="tpch_region",
    )


@pytest.fixture
def nation():
    return ibis.table(
        dict(
            n_nationkey="int32",
            n_name="string",
            n_regionkey="int32",
            n_comment="string",
        ),
        name="tpch_nation",
    )


@pytest.fixture
def customer():
    return ibis.table(
        dict(
            c_custkey="int64",
            c_name="string",
            c_address="string",
            c_nationkey="int32",
            c_phone="string",
            c_acctbal="decimal(12, 2)",
            c_mktsegment="string",
            c_comment="string",
        ),
        name="tpch_customer",
    )


@pytest.fixture
def orders():
    return ibis.table(
        dict(
            o_orderkey="int64",
            o_custkey="int64",
            o_orderstatus="string",
            o_totalprice="decimal(12, 2)",
            o_orderdate="string",
            o_orderpriority="string",
            o_clerk="string",
            o_shippriority="int32",
            o_comment="string",
        ),
        name="tpch_orders",
    )


@pytest.fixture
def tpch(region, nation, customer, orders):
    fields_of_interest = [
        customer,
        region.r_name.name("region"),
        orders.o_totalprice,
        orders.o_orderdate.cast("timestamp").name("odate"),
    ]

    return (
        region.join(nation, region.r_regionkey == nation.n_regionkey)
        .join(customer, customer.c_nationkey == nation.n_nationkey)
        .join(orders, orders.o_custkey == customer.c_custkey)[fields_of_interest]
    )


def test_join_key_name(tpch, snapshot):
    year = tpch.odate.year().name("year")

    pre_sizes = tpch.group_by(year).size()
    t2 = tpch.view()
    conditional_avg = t2[t2.region == tpch.region].o_totalprice.mean().name("mean")
    amount_filter = tpch.o_totalprice > conditional_avg
    post_sizes = tpch[amount_filter].group_by(year).size()

    percent = (post_sizes[1] / pre_sizes[1].cast("double")).name("fraction")

    expr = pre_sizes.join(post_sizes, pre_sizes.year == post_sizes.year)[
        pre_sizes.year,
        pre_sizes[1].name("pre_count"),
        post_sizes[1].name("post_count"),
        percent,
    ]
    result = ibis.impala.compile(expr)
    snapshot.assert_match(result, "out.sql")


def test_join_key_name2(tpch, snapshot):
    year = tpch.odate.year().name("year")

    pre_sizes = tpch.group_by(year).size()
    post_sizes = tpch.group_by(year).size().view()

    expr = pre_sizes.join(post_sizes, pre_sizes.year == post_sizes.year)[
        pre_sizes.year,
        pre_sizes[1].name("pre_count"),
        post_sizes[1].name("post_count"),
    ]
    result = ibis.impala.compile(expr)
    snapshot.assert_match(result, "out.sql")


def test_group_by_with_window_preserves_range(snapshot):
    t = ibis.table(dict(one="string", two="double", three="int32"), name="my_data")
    w = ibis.cumulative_window(order_by=t.one)
    expr = t.group_by(t.three).mutate(four=t.two.sum().over(w))

    result = ibis.impala.compile(expr)
    snapshot.assert_match(result, "out.sql")
