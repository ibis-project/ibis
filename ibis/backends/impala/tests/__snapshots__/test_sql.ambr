# name: test_group_by_with_window_preserves_range
  '''
  SELECT *,
         sum(`two`) OVER (PARTITION BY `three` ORDER BY `one` ASC RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS `four`
  FROM my_data
  '''
# ---
# name: test_is_parens[isnull]
  '''
  SELECT *
  FROM `table`
  WHERE (`a` IS NULL) = (`b` IS NULL)
  '''
# ---
# name: test_is_parens[notnull]
  '''
  SELECT *
  FROM `table`
  WHERE (`a` IS NOT NULL) = (`b` IS NOT NULL)
  '''
# ---
# name: test_is_parens_identical_to
  '''
  SELECT *
  FROM `table`
  WHERE (`a` IS NOT DISTINCT FROM NULL) = (`b` IS NOT DISTINCT FROM NULL)
  '''
# ---
# name: test_join_aliasing
  '''
  WITH t0 AS (
    SELECT `d`, `c`
    FROM t2
  ),
  t1 AS (
    SELECT `d`, CAST(`d` / 15 AS bigint) AS `idx`, `c`, count(1) AS `row_count`
    FROM t0
    GROUP BY 1, 2, 3
  ),
  t2 AS (
    SELECT *, `a` + 20 AS `d`
    FROM test_table
  )
  SELECT t3.*, t4.`total`
  FROM (
    SELECT `d`, `b`, count(1) AS `count`, count(DISTINCT `c`) AS `unique`
    FROM t2
    GROUP BY 1, 2
  ) t3
    INNER JOIN (
      SELECT t5.*
      FROM (
        SELECT t1.*, t8.`total`
        FROM t1
          INNER JOIN (
            SELECT `d`, sum(`row_count`) AS `total`
            FROM t1
            GROUP BY 1
          ) t8
            ON t1.`d` = t8.`d`
      ) t5
      WHERE t5.`row_count` < (t5.`total` / 2)
    ) t4
      ON t3.`d` = t4.`d`
  '''
# ---
# name: test_join_key_name
  '''
  WITH t0 AS (
    SELECT t5.*, t3.`r_name` AS `region`, t6.`o_totalprice`,
           CAST(t6.`o_orderdate` AS timestamp) AS `odate`
    FROM tpch_region t3
      INNER JOIN tpch_nation t4
        ON t3.`r_regionkey` = t4.`n_regionkey`
      INNER JOIN tpch_customer t5
        ON t5.`c_nationkey` = t4.`n_nationkey`
      INNER JOIN tpch_orders t6
        ON t6.`o_custkey` = t5.`c_custkey`
  )
  SELECT t1.`year`, t1.`count` AS `pre_count`, t2.`count` AS `post_count`,
         t2.`count` / CAST(t1.`count` AS double) AS `fraction`
  FROM (
    SELECT extract(`odate`, 'year') AS `year`, count(1) AS `count`
    FROM t0
    GROUP BY 1
  ) t1
    INNER JOIN (
      SELECT extract(t0.`odate`, 'year') AS `year`, count(1) AS `count`
      FROM t0
      WHERE t0.`o_totalprice` > (
        SELECT avg(t7.`o_totalprice`) AS `mean`
        FROM t0 t7
        WHERE t7.`region` = t0.`region`
      )
      GROUP BY 1
    ) t2
      ON t1.`year` = t2.`year`
  '''
# ---
# name: test_join_key_name2
  '''
  WITH t0 AS (
    SELECT t5.*, t3.`r_name` AS `region`, t6.`o_totalprice`,
           CAST(t6.`o_orderdate` AS timestamp) AS `odate`
    FROM tpch_region t3
      INNER JOIN tpch_nation t4
        ON t3.`r_regionkey` = t4.`n_regionkey`
      INNER JOIN tpch_customer t5
        ON t5.`c_nationkey` = t4.`n_nationkey`
      INNER JOIN tpch_orders t6
        ON t6.`o_custkey` = t5.`c_custkey`
  ),
  t1 AS (
    SELECT extract(`odate`, 'year') AS `year`, count(1) AS `count`
    FROM t0
    GROUP BY 1
  )
  SELECT t1.`year`, t1.`count` AS `pre_count`, t2.`count` AS `post_count`
  FROM t1
    INNER JOIN t1 t2
      ON t1.`year` = t2.`year`
  '''
# ---
# name: test_join_no_predicates_for_impala[cross_join]
  '''
  SELECT t0.*
  FROM star1 t0
    CROSS JOIN star2 t1
  '''
# ---
# name: test_join_no_predicates_for_impala[inner_join]
  '''
  SELECT t0.*
  FROM star1 t0
    CROSS JOIN star2 t1
  '''
# ---
# name: test_join_no_predicates_for_impala[left_join]
  '''
  SELECT t0.*
  FROM star1 t0
    CROSS JOIN star2 t1
  '''
# ---
# name: test_join_no_predicates_for_impala[outer_join]
  '''
  SELECT t0.*
  FROM star1 t0
    CROSS JOIN star2 t1
  '''
# ---
# name: test_join_with_nested_or_condition
  '''
  SELECT t0.*
  FROM t t0
    INNER JOIN t t1
      ON (t0.`a` = t1.`a`) AND
         ((t0.`a` != t1.`b`) OR (t0.`b` != t1.`a`))
  '''
# ---
# name: test_join_with_nested_xor_condition
  '''
  SELECT t0.*
  FROM t t0
    INNER JOIN t t1
      ON (t0.`a` = t1.`a`) AND
         (((t0.`a` != t1.`b`) OR (t0.`b` != t1.`a`)) AND NOT ((t0.`a` != t1.`b`) AND (t0.`b` != t1.`a`)))
  '''
# ---
# name: test_limit_cte_extract
  '''
  WITH t0 AS (
    SELECT *
    FROM functional_alltypes
    LIMIT 100
  )
  SELECT t0.*
  FROM t0
    CROSS JOIN t0 t1
  '''
# ---
# name: test_logically_negate_complex_boolean_expr
  '''
  SELECT NOT (`a` IN ('foo') AND (`c` IS NOT NULL)) AS `tmp`
  FROM t
  '''
# ---
# name: test_multiple_filters
  '''
  SELECT *
  FROM (
    SELECT *
    FROM t0
    WHERE `a` < 100
  ) t0
  WHERE `a` = (
    SELECT max(`a`) AS `max`
    FROM t0
    WHERE `a` < 100
  )
  '''
# ---
# name: test_multiple_filters2
  '''
  SELECT *
  FROM (
    SELECT *
    FROM t0
    WHERE `a` < 100
  ) t0
  WHERE (`a` = (
    SELECT max(`a`) AS `max`
    FROM t0
    WHERE `a` < 100
  )) AND
        (`b` = 'a')
  '''
# ---
# name: test_nested_join_base
  '''
  WITH t0 AS (
    SELECT `uuid`, count(1) AS `count`
    FROM t
    GROUP BY 1
  )
  SELECT t0.*
  FROM (
    SELECT `uuid`, max(`count`) AS `max_count`
    FROM t0
    GROUP BY 1
  ) t1
    LEFT OUTER JOIN t0
      ON t1.`uuid` = t0.`uuid`
  '''
# ---
# name: test_nested_join_multiple_ctes
  '''
  WITH t0 AS (
    SELECT `userid`, `movieid`, `rating`,
           CAST(`timestamp` AS timestamp) AS `datetime`
    FROM ratings
  ),
  t1 AS (
    SELECT t0.*, t5.`title`
    FROM t0
      INNER JOIN movies t5
        ON t0.`movieid` = t5.`movieid`
  )
  SELECT t2.*
  FROM (
    SELECT t1.*
    FROM t1
    WHERE (t1.`userid` = 118205) AND
          (extract(t1.`datetime`, 'year') > 2001)
  ) t2
  WHERE t2.`movieid` IN (
    SELECT `movieid`
    FROM (
      SELECT `movieid`
      FROM (
        SELECT t1.*
        FROM t1
        WHERE (t1.`userid` = 118205) AND
              (extract(t1.`datetime`, 'year') > 2001) AND
              (t1.`userid` = 118205) AND
              (extract(t1.`datetime`, 'year') < 2009)
      ) t5
    ) t4
  )
  '''
# ---
# name: test_nested_joins_single_cte
  '''
  WITH t0 AS (
    SELECT `uuid`, count(1) AS `count`
    FROM t
    GROUP BY 1
  )
  SELECT t1.*, t2.`last_visit`
  FROM (
    SELECT t0.*
    FROM (
      SELECT `uuid`, max(`count`) AS `max_count`
      FROM t0
      GROUP BY 1
    ) t3
      LEFT OUTER JOIN t0
        ON (t3.`uuid` = t0.`uuid`) AND
           (t3.`max_count` = t0.`count`)
  ) t1
    LEFT OUTER JOIN (
      SELECT `uuid`, max(`ts`) AS `last_visit`
      FROM t
      GROUP BY 1
    ) t2
      ON t1.`uuid` = t2.`uuid`
  '''
# ---
# name: test_relabel_projection
  '''
  SELECT `foo` AS `one`, `bar`, `baz` AS `three`
  FROM `table`
  '''
# ---
